gomock/internal/validate/validate.go

```diff
@@ -0,0 +1,199 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validate
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// InputAndOutputSig compares the argument and return signatures of actualFunc
+// against expectedFunc. It returns an error unless everything matches.
+func InputAndOutputSig(actualFunc, expectedFunc reflect.Type) error {
+	if err := InputSig(actualFunc, expectedFunc); err != nil {
+		return err
+	}
+
+	if err := outputSig(actualFunc, expectedFunc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// InputSig compares the argument signatures of actualFunc
+// against expectedFunc. It returns an error unless everything matches.
+func InputSig(actualFunc, expectedFunc reflect.Type) error {
+	// check number of arguments and type of each argument
+	if actualFunc.NumIn() != expectedFunc.NumIn() {
+		return fmt.Errorf(
+			"expected function to have %d arguments not %d",
+			expectedFunc.NumIn(), actualFunc.NumIn())
+	}
+
+	lastIdx := expectedFunc.NumIn()
+
+	// If the function has a variadic argument validate that one first so that
+	// we aren't checking for it while we iterate over the other args
+	if expectedFunc.IsVariadic() {
+		if ok := variadicArg(lastIdx, actualFunc, expectedFunc); !ok {
+			i := lastIdx - 1
+			return fmt.Errorf(
+				"expected function to have"+
+					" arg of type %v at position %d"+
+					" not type %v",
+				expectedFunc.In(i), i, actualFunc.In(i),
+			)
+		}
+
+		lastIdx--
+	}
+
+	for i := 0; i < lastIdx; i++ {
+		expectedArg := expectedFunc.In(i)
+		actualArg := actualFunc.In(i)
+
+		if err := arg(actualArg, expectedArg); err != nil {
+			return fmt.Errorf("input argument at %d: %s", i, err)
+		}
+	}
+
+	return nil
+}
+
+func outputSig(actualFunc, expectedFunc reflect.Type) error {
+	// check number of return vals and type of each val
+	if actualFunc.NumOut() != expectedFunc.NumOut() {
+		return fmt.Errorf(
+			"expected function to have %d return vals not %d",
+			expectedFunc.NumOut(), actualFunc.NumOut())
+	}
+
+	for i := 0; i < expectedFunc.NumOut(); i++ {
+		expectedArg := expectedFunc.Out(i)
+		actualArg := actualFunc.Out(i)
+
+		if err := arg(actualArg, expectedArg); err != nil {
+			return fmt.Errorf("return argument at %d: %s", i, err)
+		}
+	}
+
+	return nil
+}
+
+func variadicArg(lastIdx int, actualFunc, expectedFunc reflect.Type) bool {
+	if actualFunc.In(lastIdx-1) != expectedFunc.In(lastIdx-1) {
+		if actualFunc.In(lastIdx-1).Kind() != reflect.Slice {
+			return false
+		}
+
+		expectedArgT := expectedFunc.In(lastIdx - 1)
+		expectedElem := expectedArgT.Elem()
+		if expectedElem.Kind() != reflect.Interface {
+			return false
+		}
+
+		actualArgT := actualFunc.In(lastIdx - 1)
+		actualElem := actualArgT.Elem()
+
+		if ok := actualElem.ConvertibleTo(expectedElem); !ok {
+			return false
+		}
+
+	}
+
+	return true
+}
+
+func interfaceArg(actualArg, expectedArg reflect.Type) error {
+	if !actualArg.ConvertibleTo(expectedArg) {
```

Comment:
Honest question: Does this cover the implements?

Reply:
Yes it does, if expectedArg is something like `Stringer` and `actualArg` is an instance of some type `Foo` which implements `Stringer` this line will evaluate to true.

Reply:
I added another test case to make that more clear